src/platform/Darwin/DnssdImpl.cpp

```diff
@@ -340,221 +222,129 @@ void OnBrowseAdd(BrowseContext * context, const char * name, const char * type,
+                                               return strcmp(name, service.mName) == 0 && type == GetFullType(&service) &&
+                                                   service.mInterface == Inet::InterfaceId(interfaceId);
                                            }),
                             context->services.end());
 }
 
 static void OnBrowse(DNSServiceRef sdRef, DNSServiceFlags flags, uint32_t interfaceId, DNSServiceErrorType err, const char * name,
                      const char * type, const char * domain, void * context)
 {
-    BrowseContext * sdCtx = reinterpret_cast<BrowseContext *>(context);
-    VerifyOrReturn(CheckForSuccess(sdCtx, __func__, err, true));
+    auto sdCtx = reinterpret_cast<BrowseContext *>(context);
+    VerifyOrReturn(kDNSServiceErr_NoError == err, sdCtx->Finalize(err));
 
-    (flags & kDNSServiceFlagsAdd) ? OnBrowseAdd(sdCtx, name, type, domain, Inet::InterfaceId(interfaceId))
-                                  : OnBrowseRemove(sdCtx, name, type, domain, Inet::InterfaceId(interfaceId));
+    (flags & kDNSServiceFlagsAdd) ? OnBrowseAdd(sdCtx, name, type, domain, interfaceId)
+                                  : OnBrowseRemove(sdCtx, name, type, domain, interfaceId);
 
     if (!(flags & kDNSServiceFlagsMoreComing))
     {
-        sdCtx->callback(sdCtx->context, sdCtx->services.data(), sdCtx->services.size(), CHIP_NO_ERROR);
-        MdnsContexts::GetInstance().Remove(sdCtx);
+        sdCtx->Finalize();
     }
 }
 
 CHIP_ERROR Browse(void * context, DnssdBrowseCallback callback, uint32_t interfaceId, const char * type,
                   DnssdServiceProtocol protocol)
 {
-    DNSServiceErrorType err;
-    DNSServiceRef sdRef;
-    BrowseContext * sdCtx;
+    auto sdCtx = chip::Platform::New<BrowseContext>(context, callback, protocol);
+    VerifyOrReturnError(nullptr != sdCtx, CHIP_ERROR_NO_MEMORY);
 
-    std::string regtype(type);
-    std::string subtypeDelimiter = "._sub.";
-    size_t position              = regtype.find(subtypeDelimiter);
-    if (position != std::string::npos)
-    {
-        regtype = regtype.substr(position + subtypeDelimiter.size()) + "," + regtype.substr(0, position);
-    }
-
-    sdCtx = chip::Platform::New<BrowseContext>(context, callback, protocol);
-    err   = DNSServiceBrowse(&sdRef, 0 /* flags */, interfaceId, regtype.c_str(), kLocalDot, OnBrowse, sdCtx);
-    VerifyOrReturnError(CheckForSuccess(sdCtx, __func__, err), CHIP_ERROR_INTERNAL);
-
-    err = DNSServiceSetDispatchQueue(sdRef, chip::DeviceLayer::PlatformMgrImpl().GetWorkQueue());
-    VerifyOrReturnError(CheckForSuccess(sdCtx, __func__, err, true), CHIP_ERROR_INTERNAL);
+    DNSServiceRef sdRef;
+    auto err = DNSServiceBrowse(&sdRef, kBrowseFlags, interfaceId, type, kLocalDot, OnBrowse, sdCtx);
+    VerifyOrReturnError(kDNSServiceErr_NoError == err, sdCtx->Finalize(err));
 
     return MdnsContexts::GetInstance().Add(sdCtx, sdRef);
```

Comment:
Again, the error handling here seems pretty weird....

Reply:
See https://github.com/project-chip/connectedhomeip/pull/17333#discussion_r852773943